### PR TITLE
Contracts

### PR DESCRIPTION
**Pull request by cptspacecadet**

### CODE REVIEW

**Review by xBA5ED — COMMENTED** [review]
The `tokenId` generation for both `OpenTiered` and `Tiered` is 🤌🏽 *chefs kiss*

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    override
+  {
+    if (contribution.token != contributionToken) {
+      return;
+    }
+
+    uint256 tokenId;
+    for (uint256 i; i < tiers.length; i++) {
+      if (
+        (tiers[i].contributionFloor <= contribution.value && i == tiers.length - 1) ||
```
Comment (drgorillamd) [inline contracts/OpenTieredNFTRewardDelegate.sol:107]: Every tiers[i] access is another storage access

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  function validateContribution(address account, JBTokenAmount calldata contribution)
+    internal
+    override
+  {
+    if (contribution.token != contributionToken) {
+      return;
+    }
+
+    uint256 tokenId;
+    for (uint256 i; i < tiers.length; i++) {
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply contracts/OpenTieredNFTRewardDelegate.sol:105]: \+ uncheck ++i (since max 255)

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      ) {
+        tokenId = i | (uint248(uint256(keccak256(abi.encodePacked(account, block.number)))) << 8);
+        break;
+      }
+    }
+
+    if (tokenId != 0) {
+      _mint(account, tokenId);
+
+      _supply += 1;
```
Comment (drgorillamd) [inline contracts/OpenTieredNFTRewardDelegate.sol:119]: ++supply is cheaper

File: contracts/SimpleNFTRewardDelegate.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    _minContribution = minContribution;
+  }
+
+  /**
+   */
+  function validateContribution(address account, JBTokenAmount calldata contribution)
+    internal
+    override
+  {
+    if (
```
Comment (drgorillamd) [inline contracts/SimpleNFTRewardDelegate.sol:56]: 3 _minContribution sread -> 1 is enough, put it in memory plz

File: contracts/SimpleNFTRewardDelegate.sol [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      (contribution.value >= _minContribution.value &&
+        contribution.token == _minContribution.token) || _minContribution.value == 0 // TODO: should probably consider _maxSupply too
+    ) {
+      uint256 tokenId = _supply;
+      _mint(account, tokenId);
+
+      _supply += 1;
```
Comment (drgorillamd) [inline contracts/SimpleNFTRewardDelegate.sol:63]: ++supply;

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      _symbol,
+      _uri,
+      _tokenUriResolverAddress,
+      _contractMetadataUri,
+      _admin
+    )
+  {
+    contributionToken = _contributionToken;
+
+    if (_tiers.length > type(uint8).max) {
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply contracts/OpenTieredNFTRewardDelegate.sol:74]: nvm, you read it A LOT later ->`uint256 tierArrayLength`

File: contracts/SimpleNFTRewardDelegate.sol [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  /**
+   */
+  function validateContribution(address account, JBTokenAmount calldata contribution)
+    internal
+    override
+  {
+    if (
+      (contribution.value >= _minContribution.value &&
+        contribution.token == _minContribution.token) || _minContribution.value == 0 // TODO: should probably consider _maxSupply too
```
Comment (drgorillamd) [inline contracts/SimpleNFTRewardDelegate.sol:58]: if contribution.value > 0 but minContribution.value == 0, we just keep the contribution and "sorry for that guy"?

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+pragma solidity 0.8.6;
+
+import './abstract/AbstractNFTRewardDelegate.sol';
+
+/**
+  @dev Token id 0 has special meaning in NFTRewardDataSourceDelegate where minting will be skipped.
+  @dev An example tier collecting might look like this:
+  [ { contributionFloor: 1 ether }, { contributionFloor: 5 ether }, { contributionFloor: 10 ether } ]
+ */
+struct OpenRewardTier {
```
Comment (drgorillamd) [inline contracts/OpenTieredNFTRewardDelegate.sol:11]: Why the struct?

File: contracts/TieredNFTRewardDelegate.sol [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    if (totalOwnerBalance(account) >= userMintCap) {
+      return;
+    }
+
+    uint256 tokenId;
+    for (uint256 i; i < tiers.length; i++) {
+      if (
+        tiers[i].contributionFloor <= contribution.value &&
+        i == tiers.length - 1 &&
+        tiers[i].remainingAllowance != 0
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply contracts/TieredNFTRewardDelegate.sol:137]: And the trick is then to use a _current scoped outside of the loop, that you overwrite on each iteration, for additional gas golfing

File: contracts/abstract/AbstractNFTRewardDelegate.sol [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+  */
+  IToken721UriResolver private _tokenUriResolver;
+
+  /**
+    @notice
+    Contract opensea-style metadata uri.
+  */
+  string private _contractUri;
+
+  bool private transferrable;
```
Comment (drgorillamd) [inline contracts/abstract/AbstractNFTRewardDelegate.sol:99]: underscore + natspec

File: contracts/abstract/AbstractNFTRewardDelegate.sol [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+  //*********************************************************************//
+  // -------------------------- constructor ---------------------------- //
+  //*********************************************************************//
+
+  /**
+    @notice
+    Project id of the project this configuration is associated with.
+  */
+  uint256 private _projectId;
```
Comment (drgorillamd) [inline contracts/abstract/AbstractNFTRewardDelegate.sol:60]: please separate private and internal variable (+ are private really needed? Would you shadow-overwrite them in children contracts without?)

File: contracts/abstract/AbstractNFTRewardDelegate.sol [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+    Transfer tokens to an account.
+
+    @param _to The destination address.
+    @param _id NFT id to transfer.
+  */
+  function transfer(address _to, uint256 _id) public override {
+    if (!transferrable) {
+      revert NON_TRANSFERRABLE();
+    }
+    transferFrom(msg.sender, _to, _id);
```
Comment (drgorillamd) [inline contracts/abstract/AbstractNFTRewardDelegate.sol:297]: you test the transferrable twice then (which is another double sread) -> `super.transferFrom(msg.sender, _to_id);`

File: contracts/abstract/AbstractNFTRewardDelegate.sol [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+    return _ownerOf[_id] == _account;
+  }
+
+  // TODO: this will cause issues for some price resolvers
+  function mint(address _account) external override onlyOwner returns (uint256 tokenId) {
+    if (_supply == _maxSupply) {
+      revert SUPPLY_EXHAUSTED();
+    }
+
+    tokenId = _supply;
```
Comment (drgorillamd) [inline contracts/abstract/AbstractNFTRewardDelegate.sol:333]: do this assignment before the _supply==_maxSupply and then use tokenId to compare (that's another double sread currently)

File: lib/forge-std/.github/workflows/tests.yml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+name: Tests
```
Comment (drgorillamd) [inline lib/forge-std/.github/workflows/tests.yml:1]: How did you install forge-std? The whole repo shouldn't be here (instead the link to the module only)

File: package.json [hunk 15]
```diff
@@ -9,6 +9,11 @@
   },
   "version": "1.0.0",
   "license": "MIT",
+  "dependencies": {
```
Comment (drgorillamd) [inline package.json:12]: Why? Use forge as package manager (via git modules) and remove this then

File: contracts/OpenTieredNFTRewardDelegate.sol [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    uint256 tokenId;
+    for (uint256 i; i < tiers.length; i++) {
+      if (
+        (tiers[i].contributionFloor <= contribution.value && i == tiers.length - 1) ||
+        (tiers[i].contributionFloor <= contribution.value &&
+          tiers[i + 1].contributionFloor > contribution.value)
+      ) {
+        tokenId = i | (uint248(uint256(keccak256(abi.encodePacked(account, block.number)))) << 8);
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply contracts/OpenTieredNFTRewardDelegate.sol:111]: \+ should it be 'reverse lookup-able'? If not, I think you don't even need the bit shift (just overwrite 8 bits)

**Pull request merged**